### PR TITLE
Falcon: Add RoPE scaling

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

In the same spirit as #24653, adds RoPE scaling to Falcon. It also borrows a few changes from #25740, to allow for `codellama`-style scaling.

In addition to the changes above, it also adds the `max_position_embeddings` to the config attributes, needed for one of the scaling strategies.

__________________________________________

Python script to validate these changes: https://pastebin.com/SJmUpDU1
Before this PR 👉 outputs gibberish
After this PR 👉 recognizes that the super large prompt is about llama 2

### CODE REVIEW

File: src/transformers/models/falcon/configuration_falcon.py
Comment (gante) [inline src/transformers/models/falcon/configuration_falcon.py:176]: This function is copy/paste from #24653 except for this `if` block

File: src/transformers/models/falcon/modeling_falcon.py
Comment (gante) [inline src/transformers/models/falcon/modeling_falcon.py:85]: separates the logic for the creation of `self.cos_cached` and `self.sin_cached`, since these are the only bits the other scaling factors need to overwrite

Comment (Rocketknight1) [reply src/transformers/models/falcon/modeling_falcon.py:85]: Makes sense to me!

File: src/transformers/models/falcon/configuration_falcon.py
Comment (Rocketknight1) [inline src/transformers/models/falcon/configuration_falcon.py:82]: This looks like two strategies!

**Review by Rocketknight1 — APPROVED** [review]
LGTM, details seem correct when I went through it and it shouldn't have any backward compatibility implications. Still, please check with the authors before merging it!

File: src/transformers/models/falcon/configuration_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/configuration_falcon.py:81]:
```suggestion
            Dictionary containing the scaling configuration for the RoPE embeddings. Currently supports two scaling
```

File: src/transformers/models/falcon/configuration_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/configuration_falcon.py:79]: Why isn't this part of the rope scaling dict?

File: src/transformers/models/falcon/configuration_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/configuration_falcon.py:181]: The keys should probably be checked! 
```suggestion
                "`rope_scaling` must be a dictionary with with two fields, `type` and `factor`, "
```

File: src/transformers/models/falcon/configuration_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/configuration_falcon.py:188]:
```suggestion
                f"`rope_scaling`'s type field must be one of ['linear', 'dynamic'], got {rope_scaling_type}"
```

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:293]: I realise this is copied elsewhere and not from this PR - but this is an indication to me this parameter is misnamed. If no rope scaling is happening, then a rope parameter shouldn't be used here

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:87]: There should be a more explicit var name than `t` here

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:88]: Could you re-write this with classic matrix operations? Unfortunately einsum creates issues when using traced models atm :(

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:124]: Same here re the parameter 't'

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:128]: And here re einsum

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:158]: Do we want to conditionally call `.float` here too?

File: src/transformers/models/falcon/modeling_falcon.py
Comment (amyeroberts) [inline src/transformers/models/falcon/modeling_falcon.py:162]: Same comments about 't' and einsum

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this! 🤗 

Just a few implementation questions / requests.

File: src/transformers/models/falcon/configuration_falcon.py
Comment (gante) [reply src/transformers/models/falcon/configuration_falcon.py:79]: Technically this isn't a rope scaling parameter, it is a constant used to compute the embeddings. As such, I agree with the original decision to keep it separate, in the CodeLlama PR :)

Coincidently, it can also be used for scaling (i.e. by increasing it then fine-tuning it).

File: src/transformers/models/falcon/modeling_falcon.py
Comment (gante) [reply src/transformers/models/falcon/modeling_falcon.py:293]: (see comment above, I believe the two questions are related)

File: src/transformers/models/falcon/modeling_falcon.py
Comment (gante) [reply src/transformers/models/falcon/modeling_falcon.py:158]: Yes, we want to keep these auxiliary calculations in fp32 and then cast down the final results (`sin_cache` and `cos_cache`) if needed :)

**Comment by gante** [comment]
`t` and `einsum` PR comments 👉  as discussed offline, this will be fixed in a follow-up PR

I've also tested this PR against _the thing I wanted to test_, it is working correctly with and without RoPE scaling!

Merging :)

**Pull request merged**
